File: go/112.go

package main

import "github.com/astaxie/beego"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 6 star, 没写出来，需要重点掌握
// 如果当前节点是叶子节点，则检查当前的和和目标值是否相等，否则，检查左节点或右节点是否满足
func hasPathSum(root *TreeNode, sum int) bool {
	return checkSum(root, sum, 0)

}

func checkSum(node *TreeNode,  sum int, curSum int) bool {
	if node == nil {return false}
	curSum += node.Val
	if node.Left == nil && node.Right == nil && sum == curSum {
		return true
	}
	return checkSum(node.Left, sum, curSum) || checkSum(node.Right, sum, curSum)
}

File: v2/112.py





# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        # 6 star， no idea, dfs, 必须掌握
        # todo, 必须掌握
        if not root:
            return False
        rs = self.dfs(root.val, root, sum)
        return True if rs else False


    def dfs(self, current, node , sum):
        if current == sum and not node.left and not node.right:
            return True
        if node:
            left = right = None
            if node.left:
                left = self.dfs(current+node.left.val, node.left, sum)
            if node.right:
                right = self.dfs(current+node.right.val, node.right, sum)
            # 左边或者右边的和等于sum了，则True
            if left or right:
                return True